### DISCUSSION TITLE
How to get most recent successful deployment by REST

### DISCUSSION THREAD

**Discussion by acoulton** [community/community#24484] [Apps, API and Webhooks]
<p>As part of our deployment flow, before creating a new deployment I’d like to show the user a compare link and summary commit log between the current active deployment and the SHA that’s now ready to deploy.</p>
<p>However, it looks like this is actually a bit tricky to figure out through the API?</p>
<p>I don’t see an option to filter the <a href="https://developer.github.com/v3/repos/deployments/#list-deployments" rel="nofollow noopener">list deployments</a> API call (e.g. to only show successful ones) and the response to that call doesn’t itself include any info about the current status of that deployment.</p>
<p>It looks like the only way to get this info would be to call “List deployments” then iterate over the result and call “list deployment statuses” for each deployment in turn until I find one that’s successful? That doesn’t seem very efficient (though hopefully usually I’d not have to go too far back in history to find it).</p>
<p>Is there a better way to do this? I note that both the most recent successful and the status of each deployment are shown on the deployments page within github so I’m wondering if that data is actually available somewhere and I’m just not seeing it.</p>

**Comment by AndreaGriffiths11** [comment]
<p>Hi <a class="mention" href="/acoulton">@acoulton</a>,</p>
<p>Thank you for being here! Perhaps one of our users might have more insight into this, but I can’t think of a better way to reach your goal other than how you are working it thus far.</p>
<p>As far as I know, there is no way to get the computed value in the deployments payload to indicate the latest status. I’ll dig a little deeper and post any updates here.</p>
<p>Best,</p>
<p>Andrea</p>

**Comment by acoulton** [comment, marked as answer]
<p>Thanks Andrea,</p>
<p>I’ve actually realised it’s much easier to solve with GraphQL. In case it helps anyone in future, the following query does the trick:</p>
<pre><code>query($repo\_owner:String!, $repo\_name:String!, $deployment\_env:String!) {  
 repository(owner: $repo\_owner, name: $repo\_name) {  
 deployments(environments: [$deployment\_env], first: 30) {  
 edges {  
 cursor  
 node {  
 state  
 commit {  
 oid  
 }  
 }  
 }  
 }  
 }  
}
</code></pre>
<p>And then it’s just a case of finding the first node in the returned result that has state === ‘ACTIVE’.</p>
<p>Obviously this could also be done with a smaller page size and then load the next page in the unlikely event there’s loads of failed / pending / whatever deployments since the last successful one but it was simpler for me to just assume it’s not going to be more than 30 deployments ago and take a bigger page in a single request.</p>
<p>One thing that did trip me up was that at first running the GraphQL query with my app’s installation token was just producing a “Could not resolve to a Repository with the name ‘…’” error.</p>
<p>After a bit of trial and error, I realised that although the app had access to deployments and code, it didn’t have access to repository metadata.</p>
<p>This isn’t an issue with REST - you can call /repos/foo/bar/deployments etc even though the /repos/foo/bar method isn’t accessible.</p>
<p>But it seems with GraphQL it prevents the query finding the repo at all, even if you don’t ask for any metadata and just want to traverse through into the collection of deployments.</p>
<p>Granting read access to repository metadata cleared the error.</p>

**Reply by bel0v** [reply]
Hi @acoulton, thanks for the script!
Not sure if something's changed since 2019 but now I had to specify ordering for it to work correctly:
```
  deployments(environments: [$environment], first: 10, orderBy: {field: CREATED_AT, direction: DESC}) {
```

**Comment by briceburg** [comment]
I was able to get the latest ACTIVE deployment for an environment and extract its URL via something like:

```bash
gh api graphql -F owner='MY_ORG/OWNER' -F name='MY_REPO' -F environment='MY_ENVIRONMENT' -f query='
    query($name: String!, $owner: String!, $environment: String!) {
      repository(owner: $owner, name: $name) {
        deployments(environments: [$environment], first: 10, orderBy: {field: CREATED_AT, direction: DESC}) {
          edges {
            node {
              state
              latestStatus { environmentUrl }
            }
          }
        }
      }
    }
  ' | jq '.data.repository.deployments.edges[].node | select(.state == "ACTIVE")'
```

substitute MY_ORG/OWNER, MY_REPO, MY_ENVIRONMENT with your own.

in this case I only want the state and environment_url, if you want other fields check the graphql documentation for the Deployment object: https://docs.github.com/en/graphql/reference/objects#deployment

**Reply by briceburg** [reply]
to extract a particular field into a bash shell variable, you can do something like:

```bash
deploy_url=$(gh api graphql -F owner='MY_ORG/OWNER' -F name='MY_REPO' -F environment='MY_ENVIRONMENT' -f query='
    query($name: String!, $owner: String!, $environment: String!) {
      repository(owner: $owner, name: $name) {
        deployments(environments: [$environment], first: 10, orderBy: {field: CREATED_AT, direction: DESC}) {
          edges {
            node {
              state
              latestStatus { environmentUrl }
            }
          }
        }
      }
    }
  ' | jq -r '.data.repository.deployments.edges[].node | select(.state == "ACTIVE").latestStatus.environmentUrl // empty')
  
if [ -n "$deploy_url" ]; then
  echo "deploy_url: $deploy_url"
else
  echo "deploy url not found. is there an active deployment for MY_ENVIRONMENT?"
fi
```
